matrix_generation/main.cpp

```diff
@@ -0,0 +1,356 @@
+#include "DynamicMatrix.h"
+#include "IsingHamiltonian.h"
+#include<iostream>
+#include<stdio.h>
+#include"slepcsys.h"
+#include<slepceps.h>
+#include"petscsys.h"
+#include<fstream>
+#include <chrono>
+#include <thread>
+
+
+
+static char help[] = "Distributed Hamiltonian Object-Oriented Generator and Solver , by Andrew Grace.";
+
+bool loadMatrix(std::string filename, Mat * mat){
+    int numberOfNodes;
+    int thisNode;
+    MPI_Comm_rank(MPI_COMM_WORLD, &thisNode);
+    MPI_Comm_size(MPI_COMM_WORLD, &numberOfNodes);
+
+    std::string line;
+    PetscPrintf(MPI_COMM_WORLD, "Got into loadMatrix\n");
+    std::ifstream matfile(filename);
+    PetscPrintf(MPI_COMM_WORLD, "got file stream\n");
+    if(matfile.is_open()){
+        int rows, cols;
+        std::stringstream ss(line);
+        //Read it off with MPI
+        if(thisNode == 0){
+            ss.clear();
+            getline(matfile, line);
+            ss.str(line);
+            ss >> rows;
+            getline(matfile, line);
+            ss.clear();
+            ss.str(line);
+            ss >> cols;
+            
+        }
+        MPI_Bcast(&rows, 1, MPI_INT, 0, MPI_COMM_WORLD);
+        MPI_Bcast(&cols, 1, MPI_INT, 0, MPI_COMM_WORLD);
+        
+
+        std::cout << rows << std::endl;
+        std::cout << cols << std::endl;
+        MatCreate(PETSC_COMM_WORLD, mat);
+        MatSetSizes(*mat, PETSC_DECIDE, PETSC_DECIDE, rows, cols);
+        MatSetFromOptions(*mat);
+        MatSetUp(*mat);
+        MatZeroEntries(*mat);
+
+        while(getline(matfile, line) && !line.empty()){
+            int extractedRow;
+            int extractedColumn;
+            double value;
+            if(thisNode == 0){
+                ss.clear();
+                ss.str(line);
+                std::string indexStr;
+                std::string valueStr;
+                ss >> indexStr;
+                ss >> valueStr;
+                long indexRaw = std::stol(indexStr);
+                value = std::stod(valueStr);
+                extractedRow = indexRaw / cols;
+                extractedColumn = indexRaw % cols;
+            }
+            MPI_Bcast(&extractedRow, 1, MPI_INT, 0, MPI_COMM_WORLD);
+            MPI_Bcast(&extractedColumn, 1, MPI_INT, 0, MPI_COMM_WORLD);
+            MPI_Bcast(&value, 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);
```

Comment:
Why broadcast? I thought the matrices were distributed across the nodes?

Reply:
I had some old code in there that I forgot to take out, my bad. Tomorrow I will do another pull request with that out. My branch is properly sanitized and has that stuff out of there.